### PR TITLE
Added Benchmarking Library

### PR DESCRIPTION
**Pull request by wnats**
Added a library for benchmarking purposes, as requested by @pschachte.

This is similar to #304, but with @pschachte's suggestion to store the clock counter as a global variable in `cbits.c`.

This library introduces two new `semipure` procedures, `!benchmark.start` and `!benchmark.end(?time:float)`, serving as a pair of probes to measure the _CPU clock time_ elapsed between two points of a program. A utility procedure `!benchmark.time_execution(proc:{semipure}(), ?time:float)` is also introduced which times `proc`. The library will throw an error if:
- `!benchmark.start` is called twice without a `!benchmark.end(?time:float)` in between
- `!benchmark.end(?time:float)` is called when an un-ended `!benchmark.start` has not been called
- `!benchmark.time_execution(...)` is called when there is a preceding `!benchmark.start` which has not been ended

### CODE REVIEW

**Comment by jimbxb** [comment]
This is cool. I like that we can guarantee that the timer is only started and ended once in succession.

One request: can we add a utility of some kind like this?

```
def {semipure} time(proc:{semipure}(), ?time:float) {
    !start
    !proc
    !end(?time)
}
```

File: wybelibs/benchmark.wybe [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# End the benchmark clock, yielding the elapsed time in seconds
+# Can only be called after a start() has been called
+# which has not been ended by an end()
+pub def {semipure} end(?time:float) use call_source_location {
+    foreign c {impure} benchmark_end(call_source_location, ?time)
+}
+
+# Measure the running time of a semipure niladic procedure
+# Assumes any start() call has been ended
+pub def {semipure} time(proc:{semipure}(), ?time:float) {
```
Comment (jimbxb) [inline wybelibs/benchmark.wybe:23]: Now that I think about it, we probably want the `call_source_location` from wherever we call `time`. I think that means we have to manually call the `foreign` procs in here too.

File: wybelibs/benchmark.wybe [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use wybe.c_string, wybe.float
+
+# Start the benchmark clock
+# Can only be called once before being ended by end()
+pub def {semipure} start() use call_source_location {
+    foreign c {impure} benchmark_start(call_source_location)
+}
+
+# End the benchmark clock, yielding the elapsed time in seconds
+# Can only be called after a start() has been called
```
Comment (jimbxb) [inline wybelibs/benchmark.wybe:15]: Nit: can we remove the prepositions on `start` and `end`. `start` doesn't use the preposition :)

**Review by jimbxb — CHANGES_REQUESTED** [review]

File: wybelibs/benchmark.wybe (hunk 1 above)
Comment (wnats) [reply wybelibs/benchmark.wybe:23]: Good point!

File: wybelibs/benchmark.wybe (hunk 2 above)
Comment (wnats) [reply wybelibs/benchmark.wybe:15]: No worries, didn't realise that, thank for pointing it out.

**Review by jimbxb — APPROVED** [review]
LGTM!

File: wybelibs/benchmark.wybe [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# End the benchmark clock, yielding the elapsed time in seconds
+# Can only be called after start() has been called
+# which has not been ended by end()
+pub def {semipure} end(?time:float) use call_source_location {
+    foreign c {impure} benchmark_end(call_source_location, ?time)
+}
+
+# Measure the running time of a semipure niladic procedure
+# Assumes the preceding start() (if any) has been ended
+pub def {semipure} time(proc:{semipure}(), ?time:float) use call_source_location {
```
Comment (pschachte) [inline wybelibs/benchmark.wybe:23]: Could you rename this proc to `time_execution`?  I think this name is pretty ambiguous.

**Review by pschachte — CHANGES_REQUESTED** [review]
All good except one request.

**Review by pschachte — APPROVED** [review]

**Comment by pschachte** [comment]
Thank you!

**Pull request merged**